src/why3_tactic.ml

```diff
@@ -0,0 +1,292 @@
+    try
+        Why3.Whyconf.load_driver why3_main !why3_env cp.Why3.Whyconf.driver []
+    with e ->
+        Console.fatal pos "Failed to load driver for %s: %a"
+        cp.prover.prover_name
+        Why3.Exn_printer.exn_printer e
+
+(** [result pos prv tsk] return the result of a prover [prv] with the task
+    [tsk]. *)
+let result :
+    Pos.popt ->
+    Why3.Whyconf.config_prover ->
+    Why3.Task.task ->
+    Why3.Call_provers.prover_result =
+    fun pos prv tsk ->
+      let limit =
+        {
+            Why3.Call_provers.empty_limit
+            with limit_time = !prover_timeout
+        } in
+      Why3.Call_provers.wait_on_call (Why3.Driver.prove_task
+      ~limit:limit
+      ~command:prv.Why3.Whyconf.command (prover_driver pos prv) tsk)
+
+(** [answer ans] check if the answer [ans] of a prover is valid or not. *)
+let answer : Why3.Call_provers.prover_answer -> bool = fun ans ->
+    ans = Why3.Call_provers.Valid
+
+(** [call pos sp tsk] call the prover named [sp] with the task [tsk]. *)
+let call :
+    Pos.popt -> string -> Why3.Task.task -> Why3.Call_provers.prover_result =
+    fun pos sp tsk -> result pos (prover pos sp) tsk
+
+let handle prover_name ss tac ps g =
+      (* Get the goal to prove. *)
+      let (hypotheses, trm) = Proof.Goal.get_type g in
+      (* Get the default or the indicated name of the prover. *)
+      let prover_name = Option.get prover_name Timed.(!default_prover) in
+      (* Translate from lambdapi to why3 terms. *)
+      let (constants_table, hyps, why3term) =
+          translate tac.Pos.pos ps.Proof.proof_builtins (hypotheses, trm)
+      in
+      (* Create a new task that contains symbols, axioms and the goal. *)
+      let tsk = create constants_table hyps why3term in
+      (* Call the prover named [prover_name] and get the result. *)
+      let prover_result = call tac.pos prover_name tsk in
+      (* If the prover succeeds to prove the goal. *)
+      if answer prover_result.pr_answer then
+        (* Create a new axiom that represents the proved goal. *)
+        let why3_axiom = Pos.make tac.pos (get_newname ()) in
+        (* Get the meta type of the current goal (with quantified context) *)
+        let trm = Timed.(!((Proof.Goal.get_meta g).meta_type)) in
+        (* Add the axiom to the current signature. *)
+        let a = Sign.add_symbol ss.signature Const why3_axiom trm [] in
+        (* Tell the user that the goal is proved (verbose 2). *)
+        Console.out 2 "%s proved the current goal@." prover_name;
+        (* Return the variable terms of each item in the context. *)
+        let terms = List.rev_map (fun (_, (x, _)) -> Vari x) hypotheses in
+        (* Apply the instance of the axiom with context. *)
+        let instance = Basics.add_args (symb a) terms in
+        (* Return the new axiom *)
+        instance
+      else
+        Console.fatal tac.pos "%s did not found a proof@." prover_name
+(* Initilizing Why3 environment. *)
+let _ = init_env ()
```

Comment:
Indentation is not correct in part of this file (one unit = 2 spaces), but thanks it is better to have only one file! :)

Reply:
Fixed in ca566ffae8498ad4c6f70c13c27198d.

Reply:
Should I use local modules in this file ?